src/corehost/cli/comhost/clsidmap.cpp

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "comhost.h"
+#include <trace.h>
+#include <utils.h>
+#include <error_codes.h>
+
+#include <cpprest/json.h>
+using namespace web;
+
+using comhost::clsid_map_entry;
+using comhost::clsid_map;
+
+bool GetModuleHandleFromAddress(void *addr, HMODULE *hModule);
+
+namespace
+{
+    pal::dll_t get_current_module()
+    {
+        HMODULE hmod = nullptr;
+        bool res = GetModuleHandleFromAddress(&get_current_module, &hmod);
+        assert(res);
+        (void)res;
+
+        return (pal::dll_t)hmod;
+    }
+
+    HRESULT string_to_clsid(_In_ const pal::string_t &str, _Out_ CLSID &clsid)
+    {
+        // If the first character of the GUID is not '{' then COM will
+        // attempt to look up the string in the CLSID table.
+        if (str[0] == _X('{'))
+        {
+            if (SUCCEEDED(::CLSIDFromString(str.data(), &clsid)))
+                return S_OK;
+        }
+
+        return __HRESULT_FROM_WIN32(ERROR_INVALID_DATA);
+    }
+
+    clsid_map parse_stream(_Inout_ pal::istream_t &json_map_raw)
+    {
+        skip_utf8_bom(&json_map_raw);
+
+        // Parse JSON
+        json::value json_map;
+        try
+        {
+            json_map = json::value::parse(json_map_raw);
+        }
+        catch (const json::json_exception&)
+        {
+            trace::error(_X("Embedded .clsidmap format is invalid"));
+            throw HResultException{ StatusCode::InvalidConfigFile };
+        }
+
+        json::object &json_obj = json_map.as_object();
+
+        // Process JSON and construct a map
+        HRESULT hr;
+        clsid_map mapping;
+        for (std::pair<utility::string_t, json::value> &prop : json_obj)
+        {
+            CLSID clsidMaybe;
+            hr = string_to_clsid(prop.first, clsidMaybe);
+            if (FAILED(hr))
+            {
+                assert(false && "Invalid CLSID");
```

Comment:
This is a core invariant based on tooling writing `CLSID` strings _or_ a nefarious user. This is an example of a classic assert rather than some kind of state we can get into.

Reply:
OK - makes sense.